source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2058 @@
+
+        struct connection_state_change_args *args =
+            aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct connection_state_change_args));
+
+        args->socket = socket;
+        args->nw_socket = nw_socket;
+        args->allocator = nw_socket->allocator;
+        args->error = error ? nw_error_get_error_code(error) : 0;
+        args->state = state;
+        args->nw_connection = nw_connection;
+
+        s_socket_acquire_internal_ref(nw_socket);
+
+        aws_task_init(task, s_process_connection_state_changed_task, args, "ConnectionStateChangedTask");
+
+        aws_event_loop_schedule_task_now(nw_socket->event_loop, task);
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+
+    } else if (state == nw_connection_state_cancelled) {
+        // If event loop is destroyed, no io events will be proceeded. Closed the internal socket.
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+        s_socket_release_internal_ref(nw_socket);
+
+    } else {
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+    }
+}
+
+static void s_process_listener_success_task(struct aws_task *task, void *args, enum aws_task_status status) {
+    (void)status;
+    struct nw_listener_connection_args *task_args = args;
+    struct aws_allocator *allocator = task_args->allocator;
+    struct nw_socket *listener_nw_socket = task_args->nw_socket;
+
+    if (status == AWS_TASK_STATUS_RUN_READY) {
+
+        if (listener_nw_socket) {
+            aws_mutex_lock(&listener_nw_socket->synced_data.lock);
+            struct aws_socket *listener = listener_nw_socket->synced_data.base_socket;
+
+            struct aws_socket *new_socket = aws_mem_calloc(allocator, 1, sizeof(struct aws_socket));
+            struct aws_socket_options options = listener->options;
+            int error = aws_socket_init(new_socket, allocator, &options);
+            if (error) {
+                aws_mem_release(allocator, new_socket);
+                nw_release(task_args->new_connection);
+                if (listener->accept_result_fn) {
+                    listener->accept_result_fn(listener, task_args->error_code, NULL, task_args->user_data);
+                }
+            } else {
+                new_socket->io_handle.data.handle = task_args->new_connection;
+
+                new_socket->io_handle.set_queue = s_client_set_dispatch_queue;
+
+                nw_endpoint_t endpoint = nw_connection_copy_endpoint(task_args->new_connection);
+                const char *hostname = nw_endpoint_get_hostname(endpoint);
+                uint16_t port = nw_endpoint_get_port(endpoint);
+
+                if (hostname != NULL) {
+                    size_t hostname_len = strlen(hostname);
+                    size_t buffer_size = AWS_ARRAY_SIZE(new_socket->remote_endpoint.address);
+                    size_t to_copy = aws_min_size(hostname_len, buffer_size);
+                    memcpy(new_socket->remote_endpoint.address, hostname, to_copy);
```

Comment:
This does not seem safe.  It would appear to allow a host name that is not null-terminated if it's exactly the size of the array.  Are we doing this anywhere else?